### PR TITLE
feat: support conditional expressions in vectorized map

### LINKED ISSUE #4601: Vectorize conditional expressions `if/then/else`
Not sure what the concerning details are around this, but `if/then/else` was flagged as having some complications for vectorization.
The current thinking is it could be best to deliver #4599 before beginning work on this.

Notes:

Generally the conditional, consequent and alternate will need to be evaluated for all rows in the vector. However if the conditional is either all true or all false we can evaluate only one of the consequent or alternate. But if even one value of the conditional is different then we need to evaluate both the consequent and alternate for all rows and use the conditional vector to mask the values. 

DOD:
- provide a vectorized implementation for conditional expressions `if/then/else`
- guard the new implementation with a feature flag

### PR DESCRIPTION
**Pull request by onelson**
Fixes #4601

Adds support for vectorized functions containing conditional expressions.
This new code path is guarded by a feature flag called `vectorizedConditionals`, enabled in test but off by default.

In cases where the `test` (ie the condition part of the expression) is all one value or another, we can skip a lot of the new code by returning either the the `consequent` or `alternate` vector directly.

For other situations, both the `consequent` and `alternate` vectors must be evaluated in their entirety after which values are selected for return based on the `test` value for the given index.
**N.b.** this _could give an inconsistent outcome_ for cases where these expressions have some sort of side-effect when the when compared to the non-vectorized version. At this time, such expressions would prevent the function from being vectorized in the first place, but we may need to revisit this detail in the future.

The flux acceptance tests pass with or without the flag enabled, but in the future when #4739 lands, these testcases should be updated to assert the vectorized code path is actually being hit.

Go unit tests verify the behavior in a more isolated way.

### Done checklist
- [x] docs/SPEC.md updated **N/A**
- [x] Test cases written

### CODE REVIEW

File: compiler/vectorized_test.go [hunk 1]
```diff
@@ -461,6 +461,8 @@ func TestVectorizedFns(t *testing.T) {
 				if fn.Vectorized == nil {
 					t.Fatal("Expected to find vectorized node, but found none")
 				}
+				// XXX: test the vectorized version instead of the row-based one.
```
Comment (OfTheDelmer) [inline compiler/vectorized_test.go:464]: Great catch!

**Review by OfTheDelmer — COMMENTED** [review]
It looks great!  Just adding comments from our dive into `IsValid` vs `IsNull`.

**Comment by onelson** [comment]
@OfTheDelmer thanks for your notes. I'll do another pass on the null handling, hopefully with some new tests to verify that:

- when `test` is `null` it is considered `false`
- when either `consequent` or `alternate` are `null`
  - ... and they are the selected branch
  - we produce a null in the output vector

I think the blocks called out in the review may simplify further with this contract in mind.

**Comment by onelson** [comment]
Tests added in https://github.com/influxdata/flux/pull/5017/commits/2afaa998513f9a5c9f3a8596370d24054942ac0c for null inputs in each of the three vectors appear to be passing as-is, so that's something.

Working to simplify the checks now.

**Comment by onelson** [comment]
@OfTheDelmer I've simplified the null checks in https://github.com/influxdata/flux/pull/5017/commits/65ac9608784ba6852ddcd3ef65c39dda9d6b1872 -- if these changes seem right to you, can you resolve the remarks in your review?

File: array/conditional.gen.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		b.Resize(n)
+
+		if !(c.Len() == n && a.Len() == n) {
+			return nil, errors.Newf(codes.Invalid, "vectors must be equal length")
+		}
+
+		for i := 0; i < n; i++ {
+			// nulls are considered as false
+			truthy := t.IsValid(i) && t.Value(i)
+			if truthy && c.IsValid(i) {
```
Comment (OfTheDelmer) [inline array/conditional.gen.go.tmpl:22]: Nice 🚀

File: values/vector_values.gen.go [hunk 3]
```diff
@@ -81,47 +87,71 @@ func newVectorFromSlice(values []Value, typ semantic.MonoType, mem memory.Alloca
 	case semantic.BasicInt:
 		b := arrow.NewIntBuilder(mem)
 		for _, v := range values {
-			b.Append(v.Int())
+			if v.IsNull() {
```
Comment (OfTheDelmer) [inline values/vector_values.gen.go:90]: Awesome!

**Review by OfTheDelmer — APPROVED** [review]
This update looks great

**Pull request merged**